KAFKA-9330: Skip `join` when `AdminClient.close` is called in callback thread

soondenana opened the pull request:
The close method calls `Thread.join` to wait for AdminClient thread to
die, but if the close is called in the api completion callback, `join`
waits forever, as the thread calling `join` is same as the thread it
wants to wait to die.

This change checks for this condition and skips the join. The thread
will then return to main loop, where it will check for this condition
and exit.

Added a new unit test to invoke this condition. The test fails with
timeout if the thread is joined in callback, passes otherwise.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soondenana committed:
KAFKA-9330: Skip `join` when `AdminClient.close` is called in callback thread

The close method calls `Thread.join` to wait for AdminClient thread to
die, but if the close is called in the api completion callback, `join`
waits forever, as the thread calling `join` is same as the thread it
wants to wait to die.

This change checks for this condition and skips the join. The thread
will then return to main loop, where it will check for this condition
and exit.

Added a new unit test to invoke this condition. The test fails with
timeout if the thread is joined in callback, passes otherwise.

soondenana commented:
retest this please

soondenana commented:
`DelegationTokenEndToEndAuthorizationTest.testNoConsumeWithoutDescribeAclViaSubscribe` failed with error `TopicAuthorizationException: Not authorized to access topics: [topic2]`. 

The failure is unrelated to this change.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -552,11 +552,12 @@ public void close(Duration timeout) {
-
-            AppInfoParser.unregisterAppInfo(JMX_PREFIX, clientId, metrics);
-
+            // KAFKA-9330: close() can be called by AdminClient thread when it invokes
```
nit: I think there's no need to reference the JIRA. The explanation seems clear enough without any additional context.

soondenana replied:
Done.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1219,6 +1220,9 @@ public void run() {
                 handleResponses(now, responses);
             }
+
+            AppInfoParser.unregisterAppInfo(JMX_PREFIX, clientId, metrics);
```
Perhaps some of this cleanup should go in a `finally`?

soondenana replied:
Done. Created a new method for while loop and wrapped that up in try-finally.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -259,6 +260,25 @@ public void testCloseAdminClient() {
+     */
+    @Test(timeout = 5_000)
+    public void testCloseAdminClientInCallback() throws InterruptedException {
+        AdminClientUnitTestEnv env = mockClientEnv();
```
Can we change this to use MockTime instead?

soondenana replied:
Done.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -259,6 +260,25 @@ public void testCloseAdminClient() {
+     * Test if admin client can be closed in the callback invoked when
+     * an api call completes.
+     *
+     * KAFKA-9330: If calling close in callback, AdminClient thread hangs
```
nit: no need to reference jira

soondenana replied:
Removed.

hachikuji reviewed (commented):
Thanks, left a few comments.

soondenana committed:
Take care of review comments

1. Use finally to release resources
2. Use MockTime to timeout request to make sure future gets invoked.

soondenana commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -552,11 +552,12 @@ public void close(Duration timeout) {
+            // cause deadlock, so check for that condition.
+            if (Thread.currentThread() != thread) {
+                // Wait for the thread to be joined.
+                thread.join();
```
Should we be passing the close timeout to `join`?

soondenana replied:
Yes, makes sense. Fixed.

soondenana committed:
Pass close timeout to join

This makes sure that we honor the timeout passed to close

soondenana commented:
retest this please.

hachikuji reviewed (approved):
LGTM. Thanks for the patch!

hachikuji commented:
retest this please

hachikuji commented:
retest this please

hachikuji commented:
retest this please

soondenana commented:
JDK 8 build has one test faliure:
> kafka.api.GroupEndToEndAuthorizationTest.testNoConsumeWithoutDescribeAclViaSubscribe

JDK 11 build was hung and got aborted. Going to run it again.

soondenana commented:
retest this please

hachikuji commented:
retest this please
